src/business-logic/services/status-message-service-implementation.ts

```diff
@@ -0,0 +1,78 @@
+import { StatusMessage } from '../../model/entities/status-message'
+import { StatusMessageService } from './interfaces/status-message-service'
+import { StatusMessageEventEmitter } from './interfaces/status-message-event-emitter'
+import { StatusMessageRepository } from '../../data-access/repositories/interfaces/status-message-repository'
+import { StatusCode } from '../../model/enums/status-code'
+import { NotFoundException } from '../../model/exceptions/not-found-exception'
+
+export class StatusMessageServiceImplementation implements StatusMessageService {
+
+  constructor(
+    private readonly statusMessageEventEmitter: StatusMessageEventEmitter,
+    private readonly statusMessageRepository: StatusMessageRepository
+  ) { }
+
+  public async updateStatusMessages(statusMessage: StatusMessage[]): Promise<void> {
+    await Promise.all(
+      statusMessage.map(statusMessage => this.updateStatusMessage(statusMessage))
+    )
+  }
+
+  public async updateStatusMessage(statusMessage: StatusMessage): Promise<void> {
+    const statusMessageFromDatabase: StatusMessage | undefined = await this.getStatusMessageFromDatabase(statusMessage.id)
+    if (!statusMessageFromDatabase) {
+      return this.createNewStatusMessageIfStatusIsNotGood(statusMessage)
+    }
+
+    if (!this.isStatusMessagesDifferent(statusMessage, statusMessageFromDatabase)) {
+      return
+    }
+
+    this.statusMessageEventEmitter.emitStatusMessageEvent(statusMessage)
+
+    if ([StatusCode.GOOD].includes(statusMessage.statusCode)) {
+      await this.statusMessageRepository.deleteStatusMessage(statusMessage.id)
+      return
+    }
+
+    await this.statusMessageRepository.updateStatusMessage(statusMessage)
+  }
+
+  private async createNewStatusMessageIfStatusIsNotGood(statusMessage: StatusMessage): Promise<void> {
+    if ([StatusCode.GOOD].includes(statusMessage.statusCode)) {
+      return
+    }
+    this.statusMessageEventEmitter.emitStatusMessageEvent(statusMessage)
+    await this.statusMessageRepository.createStatusMessage(statusMessage)
+  }
+
+  private async getStatusMessageFromDatabase(statusMessageId: string): Promise<StatusMessage | undefined> {
+    try {
+      return this.statusMessageRepository.getStatusMessage(statusMessageId)
+    } catch (error) {
+      if (error instanceof NotFoundException) {
+        return
+      }
+      throw error
+    }
+  }
+
+  private isStatusMessagesDifferent(statusMessageOne: StatusMessage, statusMessageTwo: StatusMessage): boolean {
+    const isDifferentStatusCode: boolean = statusMessageOne.statusCode != statusMessageTwo.statusCode
+    const isDifferentMessage: boolean = statusMessageOne.message != statusMessageTwo.message
+    return isDifferentStatusCode || isDifferentMessage
```

Comment:
A suggestion: Use guards to make early returns. That way we don't have to evaluate all cases if the first check is enough to determine a difference.

Reply:
I think in this specific case it's so little that keeping the single return is preferred (atleast for me)

Reply:
No prob :) Just something to be considering in other places as well.
Any reason why `!=` is used over `!==`?

Reply:
> Any reason why != is used over !==?

No, I think that's an oversight :) 

It's been updated